credentials/tls_ext_test.go

```diff
@@ -236,3 +238,99 @@ func (s) TestTLS_CipherSuitesOverridable(t *testing.T) {
+// a server that doesn't support ALPN.
+func (s) TestTLS_DisabledALPN(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	initialVal := envconfig.EnforceALPNEnabled
+	defer func() {
+		envconfig.EnforceALPNEnabled = initialVal
+	}()
+
+	// Start a non gRPC TLS server.
+	config := &tls.Config{
+		Certificates: []tls.Certificate{serverCert},
+		NextProtos:   []string{}, // Empty list indicates ALPN is disabled.
+	}
+	listner, err := tls.Listen("tcp", ":0", config)
+	if err != nil {
+		t.Fatalf("Error starting TLS server: %v", err)
+	}
+	defer listner.Close()
+
+	tests := []struct {
+		name                   string
+		alpnEnforced           bool
+		wantErrMatchPattern    string
+		wantErrNonMatchPattern string
+	}{
+		{
+			name:                "enforced",
+			alpnEnforced:        true,
+			wantErrMatchPattern: "transport: .*missing selected ALPN property",
+		},
+		{
+			name:                   "not_enforced",
+			wantErrNonMatchPattern: "transport:",
+		},
+		{
+			name:                   "default_value",
+			wantErrNonMatchPattern: "transport:",
+			alpnEnforced:           initialVal,
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			envconfig.EnforceALPNEnabled = tc.alpnEnforced
+			// Listen to one TCP connection request.
+			go func() {
+				conn, err := listner.Accept()
+				if err != nil {
+					t.Errorf("tls.Accept failed err = %v", err)
+				} else {
+					_, _ = conn.Write([]byte("Hello, World!"))
+					_ = conn.Close()
+				}
+			}()
+
+			clientCreds := credentials.NewTLS(&tls.Config{
+				ServerName: serverName,
+				RootCAs:    certPool,
+			})
+
+			cc, err := grpc.NewClient("dns:"+listner.Addr().String(), grpc.WithTransportCredentials(clientCreds))
+			if err != nil {
+				t.Fatalf("grpc.NewClient error: %v", err)
+			}
+			defer cc.Close()
+			client := testgrpc.NewTestServiceClient(cc)
+			_, rpcErr := client.EmptyCall(ctx, &testpb.Empty{})
+
+			if gotCode := status.Code(rpcErr); gotCode != codes.Unavailable {
+				t.Errorf("EmptyCall returned unexpected code: got=%v, want=%v", gotCode, codes.Unavailable)
+			}
+
+			matchPat, err := regexp.Compile(tc.wantErrMatchPattern)
+			if err != nil {
+				t.Fatalf("Error message match pattern %q is invalid due to error: %v", tc.wantErrMatchPattern, err)
+			}
+
+			if tc.wantErrMatchPattern != "" && !matchPat.MatchString(status.Convert(rpcErr).Message()) {
+				t.Errorf("EmptyCall err = %v; want pattern match %q", rpcErr, matchPat)
+			}
+
+			nonMatchPat, err := regexp.Compile(tc.wantErrNonMatchPattern)
+			if err != nil {
+				t.Fatalf("Error message non-match pattern %q is invalid due to error: %v", tc.wantErrNonMatchPattern, err)
+			}
+
+			if tc.wantErrNonMatchPattern != "" && nonMatchPat.MatchString(status.Convert(rpcErr).Message()) {
+				t.Errorf("EmptyCall err = %v; want pattern missing %q", rpcErr, nonMatchPat)
+			}
```

Comment:
This whole test logic is quite convoluted. Can you please let me know what exactly you are trying to validate. We should try to simplify this.

Reply:
I wanted to validate that an RPC call is rejected when ALPN is enforced by the client and allowed when its not enforced. I was facing the following problems:
1. The gRPC server code automatically adds "h2" to list of supported protocols (named NextProtos), making it hard to disable ALPN. I was able to overcome this by providing a TLS listener to the gRPC server which completes the handshake and not providing any `TranasportCredentials` `ServerOption`.
2. The next problem was that when the client handshake fails, the client was retrying the requests till 10 seconds and getting a `DeadlineExceeded` status. This makes the test waste 10 seconds. I think this would require overriding the timeouts by providing a `Service Config`. This was making things complicated.
3. So I created a non gRPC server that just does a TLS handshake. Now the gRPC client would always see an error when sending it requests, but at different stages depending on if the TLS handshake succeeds. So I was trying to check if the error is a TLS handshake failure or "anything else". Since we're not wrapping any special error types for TLS failures, I was using the error message instead.

I've avoided all this now by testing the `ClientHandshake` method directly without creating a gRPC server or client. Please have another look.